app/V1Module/security/Policies/ExercisePermissionPolicy.php

```diff
@@ -56,16 +65,31 @@ public function isSubGroupNonStudentMember(Identity $identity, Exercise $exercis
             return false;
         }
 
+        if (empty($this->subgroupMembersCache[$user->getId()])) {
+            $this->subgroupMembersCache[$user->getId()] = [];
+        }
+        $subgroupCache = &$this->subgroupMembersCache[$user->getId()];
+
         /** @var Group $group */
         foreach ($exercise->getGroups() as $group) {
-            if ($group->isNonStudentMemberOfSubgroup($user)) {
-                return true;
+            if (!array_key_exists($group->getId(), $subgroupCache)) {
+                $subgroupCache[$group->getId()] = $group->isNonStudentMemberOfSubgroup($user);
             }
+            return $subgroupCache[$group->getId()];
```

Comment:
Isn't this logic a bit changed? The return is no longer in the IF, therefore the result of the first group in array is returned...

Reply:
Right, there should have been another IF.